### PR TITLE
Enable snapshot builds as optional PR checks 

### PR DESCRIPTION
**Pull request by nartal1**
This contributes to https://github.com/NVIDIA/spark-rapids/issues/5824 

This PR enables snapshot builds in premerge but keeping it as optional. i.e If the pre-merge CI fails only due to snapshot builds, then the author of the PR can still merge it  if the issue isn't related to the PR.

### CODE REVIEW

File: .github/workflows/mvn-verify-check.yml
Comment (nartal1) [inline .github/workflows/mvn-verify-check.yml:102]: This doesn't seem to work as intended. If one build fails, then it aborts all other builds in this job.

File: .github/workflows/mvn-verify-check.yml
Comment (nartal1) [inline .github/workflows/mvn-verify-check.yml:108]: @gerashegalov - This doesn't seem to work as intended.  It shows the build as PASSED everytime.

File: sql-plugin/src/main/340+/scala/com/nvidia/spark/rapids/shims/DecimalArithmeticOverrides.scala
Comment (nartal1) [inline sql-plugin/src/main/340+/scala/com/nvidia/spark/rapids/shims/DecimalArithmeticOverrides.scala:23]: This is to make spark340 build to fail. Shouldn't be checked in

File: .github/workflows/mvn-verify-check.yml
Comment (nartal1) [inline .github/workflows/mvn-verify-check.yml:27]: Please suggest a better name if this doesn't make sense.

File: .github/workflows/mvn-verify-check.yml
Comment (gerashegalov) [reply .github/workflows/mvn-verify-check.yml]: if this is the case we can simplify the PR by just using `SPARK_SHIM_VERSIONS_SNAPSHOTS_TAIL` at 
https://github.com/NVIDIA/spark-rapids/blob/058e684f9aa521d14c473ced9fb025d7001437b4/.github/workflows/mvn-verify-check.yml#L47

File: .github/workflows/mvn-verify-check.yml
Comment (gerashegalov) [reply .github/workflows/mvn-verify-check.yml:27]: maybe just neutral `get-shim-versions-from-dist` since we want to remain flexible to extract different version subsets for different purposes. Not super important as long as it works and has enough comments for each use case

File: .github/workflows/mvn-verify-check.yml
Comment (nartal1) [reply .github/workflows/mvn-verify-check.yml]: All other builds are cancelled once spark340 build fails. So i think there is more to be looked into.

Comment (gerashegalov) [reply .github/workflows/mvn-verify-check.yml]: did you try to disable fail-fast https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#handling-failures

File: dist/pom.xml
Comment (pxLi) [inline dist/pom.xml:58]: since 340 shim cannot pass the build for now, please do not add it to the snapshot list property,
as our nightly build will include all noSnapshot + snapshot build, if failed the snapshot jar would not be deployed to remote.

I would prefer to hardcode this in test itself before we could enable regular build for it

File: .github/workflows/mvn-verify-check.yml
Comment (nartal1) [reply .github/workflows/mvn-verify-check.yml]: I don;t think fail-fast is happening when I set `continue-on-error: true`, all builds come out as passed even when spark-340 is failing. Could see build error in the log

Comment (nartal1) [reply .github/workflows/mvn-verify-check.yml]: Correction: setting `fail-fast:false` explicitly seems to work.

File: sql-plugin/src/main/340+/scala/com/nvidia/spark/rapids/shims/DecimalArithmeticOverrides.scala
Comment (nartal1) [reply sql-plugin/src/main/340+/scala/com/nvidia/spark/rapids/shims/DecimalArithmeticOverrides.scala]: Reverting this change.

File: .github/workflows/mvn-verify-check.yml
Comment (nartal1) [reply .github/workflows/mvn-verify-check.yml]: Thanks @gerashegalov for the suggestion. I moved this matrix job and included snapshot versions in existing package-aggregator. Removed 340 from pom file and hard coding it for test.

**Comment by nartal1** [comment]
@pxLi - I am not sure what's the best way to test it. Currently Spark340 build is passing so we don't see "fail" for that build.
As you had mentioned earlier that `mvn[compile,RAT,scalastyle,docgen]` are optional builds and we should be able to merge the PR even if those fail, is it okay to merge this PR(once approved) ?

File: dist/pom.xml
Comment (nartal1) [reply dist/pom.xml]: Thanks @pxLi for letting me know the issue if we add 340 here. I have removed it from pom and added in yml file. PTAL

**Comment by nartal1** [comment]
build

**Review by gerashegalov — APPROVED** [review]
LGTM

**Review by pxLi — APPROVED** [review]

**Pull request merged**
